tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRepository.scala

```diff
@@ -0,0 +1,90 @@
+import org.apache.commons.configuration2.Configuration
+import org.apache.james.core.Username
+import org.apache.james.jmap.api.model.Size.Size
+import org.apache.james.jmap.api.model.{IdentityId, UploadId}
+import org.apache.james.jmap.api.upload.UploadRepository
+import org.apache.james.jmap.core.AccountId
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import jakarta.inject.{Inject, Named}
+
+import scala.jdk.CollectionConverters._
+
+trait PublicAssetRepository {
+  def create(username: Username, creationRequest: PublicAssetCreationRequest): Publisher[PublicAsset]
+
+  def update(username: Username, id: PublicAssetId, identityIds: Set[IdentityId]): Publisher[Void]
+
+  def revoke(username: Username, id: PublicAssetId): Publisher[Void]
+
+  def revoke(username: Username): Publisher[Void]
+
+  def get(username: Username, ids: Set[PublicAssetId]): Publisher[PublicAsset]
+
+  def list(username: Username): Publisher[PublicAsset]
+
+}
+
+class PublicAssetService @Inject()(uploadRepository: UploadRepository,
+                                   @Named("jmap") jmapConfiguration: Configuration) {
+  private val publicURIPrefix: String = jmapConfiguration.getString("publicAsset.publicURIPrefix", "")
+
+  def retrieveMetaData(id: UploadId, username: Username): SMono[(Size, ImageContentType)] =
+    SFlux(uploadRepository.listUploads(username))
+      .filter(_.uploadId.equals(id))
+      .next()
+      .map(uploadMetaData => (uploadMetaData.size, ImageContentType.from(uploadMetaData.contentType)))
+
+  def generatePublicURI(publicAssetId: PublicAssetId, accountId: AccountId): PublicURI =
+    PublicURI.from(publicAssetId, accountId, publicURIPrefix)
+
+  def generatePublicAsset(uploadId: UploadId, username: Username): SMono[PublicAsset] =
+    SFlux(uploadRepository.listUploads(username))
+      .filter(_.uploadId.equals(uploadId))
+      .next()
+      .map(uploadMetadata => {
+        val publicAssetId: PublicAssetId = PublicAssetIdFactory.generate()
+        val publicURI: PublicURI = generatePublicURI(publicAssetId, AccountId.from(username).fold(e => throw e, identity))
+        PublicAsset(
+          id = publicAssetId,
+          publicURI = publicURI,
+          size = uploadMetadata.size,
+          contentType = ImageContentType.from(uploadMetadata.contentType),
+          blobId = uploadMetadata.blobId)
+      })
+}
+
+class MemoryPublicAssetRepository()(val publicAssetService: PublicAssetService) extends PublicAssetRepository {
+  private val tableStore: Table[Username, PublicAssetId, PublicAsset] = Tables.synchronizedTable(HashBasedTable.create())
+
+  override def create(username: Username, creationRequest: PublicAssetCreationRequest): SMono[PublicAsset] =
+    publicAssetService.generatePublicAsset(creationRequest.uploadId, username)
+      .doOnNext(publicAsset => tableStore.put(username, publicAsset.id, publicAsset))
```

Comment:
From what I discussed with @Arsnael this morning, we think the `generatePublicAsset` or destroy the asset blob should be the responsibility of PublicAsset/set and not the Repository API.

Thoughts?